src/main/kotlin/uk/gov/justice/digital/hmpps/prisonercontactregistry/service/PrisonerContactRegistryService.kt

```diff
@@ -45,6 +56,36 @@ class PrisonerContactRegistryService(private val prisonApiClient: PrisonApiClien
     return contacts
   }
 
+  @Throws(VisitorNotFoundException::class, DateRangeNotFoundException::class)
+  fun getBannedDateRangeForPrisonerContacts(
+    prisonerId: String,
+    visitorIds: List<Long>,
+    fromDate: LocalDate,
+    toDate: LocalDate,
+  ): DateRangeDto {
+    val dateRange = DateRangeDto(fromDate, toDate)
+
+    val visitorBanRestrictions = getVisitorBanRestrictions(prisonerId, visitorIds)
+
+    for (restriction in visitorBanRestrictions) {
+      restriction.expiryDate?.let { expiryDate ->
+        if (expiryDate.isAfter(dateRange.toDate)) {
+          throw DateRangeNotFoundException(message = "Found visitor with restriction of 'BAN' with expiry date after our endDate, no date range possible")
+        }
+
+        if (expiryDate.isAfter(dateRange.fromDate)) {
+          dateRange.fromDate = expiryDate
+        }
+      } ?: run {
+        // If an expiry date is found to be null, it is classed as an "open-ended" ban. Thus, no suitable date range can be given.
+        // Replace with new DateRangeNotFoundException.
+        throw DateRangeNotFoundException(message = "Found visitor with restriction of 'BAN' with no expiry date, no date range possible")
```

Comment:
should it throw an Exception if expiry date is null. Should it return the passed in fromDate and toDate as dateRange if expiryDate is null.

Reply:
if expiry date is null, they have an on going ban with no end, so we can't show them any appropriate dates, as none can be returned. So i've returned an exception of DateRangeNotFound for this

Reply:
I think an exception makes sense,... because if there is no end to the ban then there is no date range. This will have to be tested throughly

Reply:
yes that seems to be right. Basically we throw a 404 if the date range is not applicable for the visitors.